pkg/controller/machineset.go

```diff
@@ -683,6 +700,75 @@ func (c *controller) terminateMachines(inactiveMachines []*v1alpha1.Machine, mac
 	return nil
 }
 
+func (c *controller) terminateStaleMachines(inactiveMachines []*v1alpha1.Machine, deletionWindow int, machineSet *v1alpha1.MachineSet) error {
+	var (
+		wg    sync.WaitGroup
+		errCh = make(chan error, deletionWindow)
+	)
+	defer close(errCh)
+
+	wg.Add(deletionWindow)
+	for i := 0; i < deletionWindow; i++ {
+		go c.prepareMachineForDeletion(inactiveMachines[i], machineSet, &wg, errCh)
+	}
+	wg.Wait()
+
+	select {
+	case err := <-errCh:
+		// all errors have been reported before and they're likely to be the same, so we'll only return the first one we hit.
+		if err != nil {
+			return err
+		}
+	default:
+	}
+
+	return nil
+}
+
+// calculateStaleMachineDeletionWindow computes an amount of Failed Machines that can be deleted based on FailedMachineDeletionRatio
+func (c *controller) calculateStaleMachineDeletionWindow(activeMachines, runningMachines, inactiveMachines []*v1alpha1.Machine, machineSet *v1alpha1.MachineSet) int {
+	var (
+		deletionWindow           int
+		numOfActiveMachines      = len(activeMachines)
+		numOfRunningMachines     = len(runningMachines)
+		numOfInactiveMachines    = len(inactiveMachines)
+		desiredMachineSetReplica = int(machineSet.Spec.Replicas)
+	)
+
+	// nothing to do
+	if numOfInactiveMachines == 0 {
+		return 0
+	}
+
+	// Will not delete any stale machines. It is expected that the user deletes these machines manually.
+	if c.safetyOptions.FailedMachineDeletionRatio == 0 {
+		return 0
+	}
+
+	if c.safetyOptions.FailedMachineDeletionRatio == 1 {
+		return numOfInactiveMachines
+	}
+
+	if desiredMachineSetReplica == 0 {
+		deletionWindow = numOfInactiveMachines
+	} else {
+		deletionWindow = int(math.Round(float64(desiredMachineSetReplica) * c.safetyOptions.FailedMachineDeletionRatio))
```

Comment:
could this be more than `numOfInactiveMachines`? i think the code will panic in `terminateStaleMachines()` if that happens. Can you please verify. thanks.

Reply:
Nice catch. I've [clamped](https://github.com/gardener/machine-controller-manager/pull/482/files#diff-16f7b87f40829e0c81cb6c486f3997ccR710-R712) it down to the deletion window.

Reply:
should we also adjust the same when throttling machine creation in `manageReplicas()`?

Reply:
After some deliberation, I fully agree. There is no reason to burst Machine creation if there is no need. Pushed the change.